File: 20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#
class Solution:
    def isValid(self, s):
        bracketsMap = {
            '(': ')',
            '[': ']',
            '{': '}'
        }
        stack = []
        for v in s:
            if v in '([{':
                stack.append(v)
            elif len(stack) > 0 and bracketsMap[stack[-1]] == v:
                stack.pop()
            else:
                return False
        return len(stack) == 0

# s = Solution()
# print('[', s.isValid('['))
# print(']', s.isValid(']'))
# print('(])', s.isValid('(])'))
# print('()', s.isValid('()'))
# print('()', s.isValid('()[]{}'))
# print('([)]', s.isValid('([)]'))
# print('{[]}', s.isValid('{[]}'))

File: 20.valid-parentheses.js
/*
 * @lc app=leetcode id=20 lang=javascript
 *
 * [20] Valid Parentheses
 */
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
  const bracketsMap = {
    '(': ')',
    '[': ']',
    '{': '}'
  }
const lens = s.length
let mid = Math.floor(lens / 2)
for(let i = 0; i < lens; i++) {
  console.log('i', bracketsMap[s[i]])
  if (bracketsMap[s[i]] !== s[lens - 1 - i] || bracketsMap[s[i]] !== s[i + 1]) {
    return false
  }
  console.log('mid', mid)
  if (i + 1 >= mid) {
    return true
  }
}
return true
};

// console.log('()', isValid('()'))
console.log('()[]{}', isValid('()[]{}'))
// console.log('([)]', isValid('([)]'))
// console.log('{[]}', isValid('{[]}'))

